### PR TITLE
Refactor api 3 d

### PR DESCRIPTION
**Pull request by jake4844**

### CODE REVIEW

File: topaz/utils/data/memory_mapped_data.py [hunk 1]
```diff
@@ -1,213 +1,406 @@
...
+
+    Args:
+        worker_id (int): ID of the worker.
+    """
+    worker_info = torch.utils.data.get_worker_info()
+    dataset = worker_info.dataset
+    worker_id = worker_info.id
+    dataset.rng = np.random.default_rng(np.random.MT19937(np.random.SeedSequence(worker_id)))
+
+def train_model(model, train_dataset, test_dataset, optimizer, criterion, num_epochs, batch_size, use_cuda):
```
Comment (tbepler) [inline topaz/utils/data/memory_mapped_data.py:348]: @jake4844 is this function here for testing?

File: topaz/training.py [hunk 2]
```diff
@@ -35,160 +35,232 @@
...
+    Returns:
+        List of lists containing grouped image paths.
+    """
     if os.path.isdir(images_path):
-        glob_base = images_path + os.sep + '*' # only get mrc files, need the header
-        image_paths = glob.glob(glob_base+'.mrc')# + glob.glob(glob_base+'.tiff') + glob.glob(glob_base+'.png')
-        image_name = [os.path.splitext(os.path.basename(x))[0] for x in image_paths]
+        glob_base = images_path + os.sep + '*'
+        image_paths = glob.glob(glob_base + '.mrc')
+        image_name = [os.path.basename(x) for x in image_paths]
```
Comment (DarnellGranberry) [inline topaz/training.py:100]: Conflict here around naming, "image.mrc" vs "image" in dataframe

File: topaz/extract.py [hunk 3]
```diff
@@ -5,69 +5,188 @@
...
+
+    Yields:
+        tuple: Tuple containing name, suppressed score, and coordinates.
+    """
     process = NonMaximumSuppression(radius, threshold, dims=dims)
     if pool is not None:
-        for name,score,coords in pool.imap_unordered(process, scores):
-            yield name,score,coords
+        for name, score in scores:
+            print(f"Debug: Initial score shape: {score.shape}")
```
Comment (tbepler) [inline topaz/extract.py:78]: TODO - remove debug print statements

File: topaz/extract.py [hunk 4]
```diff
@@ -5,69 +5,188 @@
...
+            original_shape = score.shape
+            if score.ndim != 3:
+                print(f"Warning: Expected 3D score array, but got shape {score.shape}. Attempting to reshape...")
```
Comment (tbepler) [inline topaz/extract.py:81]: TODO - remove warning or non-3d inputs

File: topaz/extract.py [hunk 5]
```diff
@@ -5,69 +5,188 @@
...
+                        chunk_coords[:, 1] += y  # y coordinate
+                        chunk_coords[:, 2] += z  # z coordinate
+                        
+                        coords_list.append(chunk_coords)
+                        scores_list.append(chunk_score)
+            
+            coords = np.concatenate(coords_list, axis=0) if coords_list else np.array([])
+            score = np.concatenate(scores_list, axis=0) if scores_list else np.array([])
+            
+            print(f"Debug: Final score shape: {score.shape}, coords shape: {coords.shape}")
```
Comment (tbepler) [inline topaz/extract.py:112]: Score patching code is here, a few TODOs:

- Support 2d input
- Make patching optional
- Add padding to each patch to remove edge artifacts
- Add patching below to else when not using a process pool
- Patching _could_ be pushed down to the score layer so that the scoring network is also applied patch-by-patch and pipelined directly into the patched NMS algorithm, but might not be worth the trouble

File: topaz/extract.py [hunk 6]
```diff
@@ -115,134 +248,320 @@ def extract_auprc(targets:np.ndarray, scores:np.ndarray, radius:float, threshold
...
+                            if use_cuda:
+                                torch.cuda.empty_cache()
+                            pbar.update(1)
+            
+            # Normalize the scores
+            scores /= np.maximum(weights, 1e-10)
+            
+            end_time = time.time()
+            print(f"Tomogram processing completed in {end_time - start_time:.2f} seconds")
+            yield path, scores, original_shape
```
Comment (tbepler) [inline topaz/extract.py:441]: Todo - make score patching optional (probably tied with extract patching) and push score patching into the `topaz.predict` module

File: topaz/training.py [hunk 7]
```diff
@@ -259,50 +360,68 @@ def report_data_stats_old(train_images, train_targets, test_images, test_targets
...
-            header = parse_header(header_bytes)
+                header = parse_header(header_bytes)
             source_total_regions += header.nz * header.ny * header.nx
-            # Read image's positives from targets
-            image_name = os.path.splitext(os.path.basename(path))[0]
-            target = targets[targets['image_name'] == image_name] # name must have no extension
-            # target = targets[targets['image_name'].str.contains(image_name)]
-            source_positive_regions += (len(target)*pixels_per_particle) # all pixels, not just center
-        # Calculate positive fraction and report
+            image_name = os.path.basename(path)
```
Comment (DarnellGranberry) [inline topaz/training.py:400]: Same naming convention issue

File: topaz/training.py [hunk 8]
```diff
@@ -442,69 +600,82 @@ def make_data_iterators_old(train_images:List[List[Union[Image.Image,np.ndarray]
...
+
     def __getitem__(self, i):
         path = self.image_paths[i]
-        # load the entire image
         img = load_image(path, make_image=False, return_header=False)
         img = torch.from_numpy(img.copy())
-        # create the target's binary mask
-        img_name = os.path.splitext(path.split('/')[-1])[0]
-        # image_name_matches = self.targets['image_name'].str.contains(img_name)
+        img_name = os.path.basename(path)
```
Comment (DarnellGranberry) [inline topaz/training.py:639]: Naming convention

File: topaz/training.py [hunk 9]
```diff
@@ -566,147 +781,247 @@ def evaluate_model(classifier, criteria, data_iterator, use_cuda=False):
...
-def fit_epochs(classifier, criteria, step_method, train_iterator, test_iterator, num_epochs
-              , save_prefix=None, use_cuda=False, output=sys.stdout):
-    ## fit the model, report train/test stats, save model if required
+        pbar.set_postfix({'loss': metrics[0]})  # Assuming the first metric is loss
+
+    # Calculate AUPRC for the entire epoch
+    auprc = average_precision(np.array(all_labels), np.array(all_scores))
+    return it, auprc
+
+def fit_epochs(classifier, criteria, step_method, train_iterator, test_iterator, num_epochs,
```
Comment (DarnellGranberry) [inline topaz/training.py:847]: What does "set" refer to?

File: topaz/training.py [hunk 10]
```diff
@@ -566,147 +781,247 @@ def evaluate_model(classifier, criteria, data_iterator, use_cuda=False):
...
+    train_iterator, test_iterator = make_data_iterators_old(train_images, train_targets, test_images, test_targets,
+                                                            classifier.width, split, args, dims=dims, to_tensor=to_tensor)
+
     fit_epochs(classifier, criteria, trainer, train_iterator, test_iterator, args.num_epochs,
                save_prefix=save_prefix, use_cuda=use_cuda, output=output)
 
     return classifier
 
+def train_model(classifier, train_images_path: str, train_targets_path: str, test_images_path: str, test_targets_path: str, use_cuda: bool,
+                save_prefix: str, output, args, dims: int = 2, num_sets: int = 1):
```
Comment (DarnellGranberry) [inline topaz/training.py:964]: num_sets again unclear
